mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl

```diff
@@ -0,0 +1,88 @@
+bucket := d.Get("bucket").(string)
+name := d.Get("name").(string)
+
+var listError, deleteObjectError error
+for deleteObjectError == nil {
+    res, err := config.NewStorageClient(userAgent).Objects.List(bucket).Prefix(name).Do()
+    if err != nil {
+        log.Printf("Error listing contents of folder %s: %v", bucket, err)
+        listError = err
+        break
+    }
+
+    if len(res.Items) == 0 {
+        break // 0 items, folder empty
+    }
+
+    if !d.Get("force_destroy").(bool) {
+        deleteErr := fmt.Errorf("Error trying to delete folder %s containing objects without force_destroy set to true", bucket)
+        log.Printf("Error! %s : %s\n\n", bucket, deleteErr)
+        return deleteErr
+    }
+    // GCS requires that a folder be empty (have no objects or object
+    // versions) before it can be deleted.
+    log.Printf("[DEBUG] GCS Folder attempting to forceDestroy\n\n")
+
+    // Create a workerpool for parallel deletion of resources. In the
+    // future, it would be great to expose Terraform's global parallelism
+    // flag here, but that's currently reserved for core use. Testing
+    // shows that NumCPUs-1 is the most performant on average networks.
+    //
+    // The challenge with making this user-configurable is that the
+    // configuration would reside in the Terraform configuration file,
+    // decreasing its portability. Ideally we'd want this to connect to
+    // Terraform's top-level -parallelism flag, but that's not plumbed nor
+    // is it scheduled to be plumbed to individual providers.
+    wp := workerpool.New(runtime.NumCPU() - 1)
+
+    for _, object := range res.Items {
+        log.Printf("[DEBUG] Found %s", object.Name)
+        object := object
+
+        wp.Submit(func() {
+            log.Printf("[TRACE] Attempting to delete %s", object.Name)
+            if err := config.NewStorageClient(userAgent).Objects.Delete(bucket, object.Name).Generation(object.Generation).Do(); err != nil {
+                deleteObjectError = err
+                log.Printf("[ERR] Failed to delete storage object %s: %s", object.Name, err)
+            } else {
+                log.Printf("[TRACE] Successfully deleted %s", object.Name)
+            }
+        })
+    }
+
+    // Wait for everything to finish.
+    wp.StopWait()
+}
+
+if listError != nil {
+    return fmt.Errorf("could not delete non-empty folder due to error when listing contents: %v", listError)
```

Comment:
Sorry if I missed this: the bucket implementation appears to attempt the bucket deletion even if there is a listError, and then only surface the error if the deletion was not successful. Was there a reason we didn't do the same here?

Reply:
in bucket deletion, we are not attempting to delete the bucket if the listError occurs

Reply:
I think we do try to delete bucket even if the listError occurs,
[Here](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/storage/resource_storage_bucket.go.tmpl#L993) we are only breaking object deletion loop if error occurs. We try to delete bucket regardless of object deletion, [Reference](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/services/storage/resource_storage_bucket.go.tmpl#L1055).

Reply:
updated logic, attempt to delete the folder after listError

Reply:
It looks like this was added to the bucket implementation here: https://github.com/GoogleCloudPlatform/magic-modules/pull/2755. You can read the PR description, but it appears this was added to get around permissions (a user may be able to delete the bucket, but not list buckets).

I would defer to the two of you if the same issue will be present for folders.

Reply:
updated the logic to attempt delete folder

Reply:
Discussed offline, LGTM!